quinn-proto/src/connection/streams/recv.rs

```diff
@@ -226,37 +195,105 @@ impl Recv {
     }
 }
 
-pub(crate) type ReadResult<T> = Result<Option<DidRead<T>>, ReadError>;
-
-/// Result of a `Streams::read` call in case the stream had not ended yet
-#[derive(Debug, Eq, PartialEq, Copy, Clone)]
-#[must_use = "A frame might need to be enqueued"]
-pub(crate) struct DidRead<T> {
-    pub result: T,
-    pub max_stream_data: ShouldTransmit,
-    pub max_data: ShouldTransmit,
+/// An iterator-like interface to read data from a receive stream
+pub struct Chunks<'a> {
+    id: StreamId,
+    streams: &'a mut Streams,
+    pending: &'a mut Retransmits,
+    recv: Option<Recv>,
+    ordered: bool,
+    read: u64,
+    blocked: bool,
 }
 
-pub(super) type StreamReadResult<T> = Result<Option<T>, ReadError>;
+impl<'a> Chunks<'a> {
+    pub(crate) fn new(
+        id: StreamId,
+        streams: &'a mut Streams,
+        pending: &'a mut Retransmits,
+        ordered: bool,
+    ) -> Result<Self, ReadError> {
+        let mut entry = match streams.recv.entry(id) {
+            Entry::Occupied(entry) => entry,
+            Entry::Vacant(_) => return Err(ReadError::UnknownStream),
+        };
 
-pub(crate) trait BytesRead {
-    fn bytes_read(&self) -> u64;
-}
+        let rs = entry.get_mut();
+        let recv = if !rs.assembler.can_read(ordered)? {
+            match rs.read_blocked() {
+                Ok(()) => {
+                    entry.remove_entry();
```

Comment:
Let's use `Entry::remove` consistently here and below, since we never need to get the ID from it.

Reply:
Scrapped.